error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/OptionalRules.java

```diff
@@ -220,6 +222,22 @@ T after(Optional<T> o1, Optional<T> o2) {
     }
   }
 
+  /**
+   * Prefer {@link Optional#orElseGet(Supplier)} over {@link Optional#orElse(Object)} if the given
+   * value is not a compile-time constant.
+   */
+  abstract static class OrElseToOrElseGet<T> {
+    @BeforeTemplate
+    T before(Optional<T> o, @NotMatches(CompileTimeConstantExpressionMatcher.class) T value) {
```

Comment:
So it seems that perhaps `CompileTimeConstantExpressionMatcher` is a bit too strict. When we run `apply-error-prone-suggestions.sh` this diff is also generated:
```diff
diff --git a/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java b/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java
index 70282dd4..5b430448 100644
--- a/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java
+++ b/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/ConflictDetectionTest.java
@@ -67,7 +67,7 @@ final class ConflictDetectionTest {
       return ConflictDetection.findMethodRenameBlocker(
               ASTHelpers.getSymbol(tree), tree.getName() + "t", state)
           .map(blocker -> buildDescription(tree).setMessage(blocker).build())
-          .orElse(Description.NO_MATCH);
+          .orElseGet(() -> Description.NO_MATCH);
     }
   }
 }
```

Similarly, as-is the rule triggers the following violation:
```
[WARN] /home/sschroevers/workspace/picnic/error-prone-support/refaster-support/src/main/java/tech/picnic/errorprone/refaster/AnnotatedCompositeCodeTransformer.java:[124,80] [Refaster Rule] OptionalRules.OrElseToOrElseGet: Refactoring opportunity
    (see https://error-prone.picnic.tech/refasterrules/OptionalRules#OrElseToOrElseGet)
  Did you mean 'return getAnnotationValue(Severity.class, Severity::value, delegate).orElseGet(()->SUGGESTION);'?
```

We might need a separate `Matcher` that allows any constant and field dereference :thinking:

Reply:
Agreed, after running the script I did notice a lot of changes and some seemed unintended, in the end we're looking to prevent potentially expensive method invocations used with `orElse`, perhaps it would be a better to match based on method invocations instead?